contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1426 @@
+    ///         called only by the delegation owner and operator.
+    function unstakeKeep(address _operator)
+        external
+        override
+        onlyOwnerOrOperator(_operator)
+    {
+        OperatorInfo storage operator = operators[_operator];
+        require(operator.keepInTStake != 0, "Nothing to unstake");
+        require(
+            getMinStaked(_operator, StakingProvider.KEEP) == 0,
+            "At least one application prevents from unstaking"
+        );
+        emit Unstaked(_operator, operator.keepInTStake);
+        operator.keepInTStake = 0;
+    }
+
+    /// @notice Reduces cached legacy NU stake amount by the provided amount.
+    ///         Reverts if there is at least one authorization higher than the
+    ///         sum of remaining legacy NU stake and liquid T stake for that
+    ///         operator or if the untaked amount is higher than the cached
+    ///         legacy stake amount. If succeeded, the legacy NU stake can be
+    ///         partially or fully undelegated on the legacy staking contract.
+    ///         This function allows to unstake from NU staking contract and
+    ///         still being able to operate in T network and earning rewards
+    ///         based on the liquid T staked. Can be called only by the
+    ///         delegation owner and operator.
+    function unstakeNu(address _operator, uint96 _amount)
+        external
+        override
+        onlyOwnerOrOperator(_operator)
+    {
+        OperatorInfo storage operator = operators[_operator];
+        (, uint96 tRemainder) = tToNu(_amount);
+        _amount -= tRemainder;
+        require(
+            _amount > 0 &&
+                _amount + getMinStaked(_operator, StakingProvider.NU) <=
+                operator.nuInTStake,
+            "Can't unstake specified amount of tokens"
+        );
+        operator.nuInTStake -= _amount;
+
+        emit Unstaked(_operator, _amount);
+    }
+
+    /// @notice Sets cached legacy stake amount to 0, sets the liquid T stake
+    ///         amount to 0 and withdraws all liquid T from the stake to the
+    ///         owner. Reverts if there is at least one non-zero authorization.
+    ///         Can be called only by the delegation owner and operator.
+    function unstakeAll(address _operator)
+        external
+        override
+        onlyOwnerOrOperator(_operator)
+    {
+        OperatorInfo storage operator = operators[_operator];
+        require(
+            operator.authorizedApplications.length == 0,
+            "At least one application is still authorized"
+        );
+        require(
+            operator.tStake == 0 ||
+                minTStakeAmount == 0 ||
+                operator.startTStakingTimestamp + MIN_STAKE_TIME <=
+                /* solhint-disable-next-line not-rely-on-time */
+                block.timestamp,
+            "Unstaking is possible only after 24 hours"
+        );
```

Comment:
I don't think it's necessary. As you said not necessary for Keep. For NU side: it will not possible at all to create new stake. For old stakers - they can choose operator once and then to unlock NU with increasing and decreasing authorization. so this will prevent from frontrunning